docs/path-resolution.rst

```diff
@@ -0,0 +1,690 @@
+.. _path-resolution:
+
+***************
+Path Resolution
+***************
+
+In order to be able to support reproducible builds on all platforms, the Solidity compiler has to
+abstract away the details of the filesystem where source files are stored.
+Paths used in imports must work the same way everywhere while the command-line interface must be
+able to work with platform-specific paths to provide good user experience.
+This section aims to explain in detail how Solidity reconciles these requirements.
+
+.. index:: ! virtual filesystem, ! source unit name
+.. _virtual-filesystem:
+
+Virtual Filesystem
+==================
+
+The paths used in :ref:`imports <import>` in a general case do not have to be filesystem paths.
+The compiler maintains an internal database (virtual filesystem) where each compiled source unit is
+assigned a unique *source unit name* which is an opaque and unstructured identifier.
+
+Source files can be placed in the virtual filesystem directly, using :ref:`Standard JSON interface
+<compiler-api>` and in this case source unit names can be anything.
+In most cases, however, the project files reside in the local filesystem and it is desirable for
+the compiler to be able to locate and load them automatically.
+To support this, the compiler passes all source unit names not found in the virtual filesystem to
```

Comment:
I think we should properly introduce the concept of "file loader" instead of just mentioning it in this sentence.

Reply:
This was a great idea. I added a section and introduced separate terms for the callback (*import callback*) and the built-in loader (*Host Filesystem Loader*). I ended up rewriting the rest of text a bit to match but I think that made it more consistent and now I don't have to mention the JS interface at every step.